analyzers/src/SonarAnalyzer.Common/Rules/IndexOfCheckAgainstZeroBase.cs

```diff
@@ -0,0 +1,92 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class IndexOfCheckAgainstZeroBase<TSyntaxKind, TExpressionSyntax, TBinaryExpressionSyntax> : SonarDiagnosticAnalyzer
+        where TSyntaxKind : struct
+        where TExpressionSyntax : SyntaxNode
+        where TBinaryExpressionSyntax : SyntaxNode
+    {
+        internal static readonly ImmutableArray<KnownType> CheckedTypes =
+            ImmutableArray.Create(
+                KnownType.System_Array,
+                KnownType.System_Collections_Generic_IList_T,
+                KnownType.System_String,
+                KnownType.System_Collections_IList
+            );
+
+        protected const string DiagnosticId = "S2692";
+        private const string MessageFormat = "0 is a valid index, but this check ignores it.";
+
+        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+        protected abstract TSyntaxKind LessThanExpression { get; }
+        protected abstract TSyntaxKind GreaterThanExpression { get; }
+
+        protected abstract bool TryGetConstantIntValue(TExpressionSyntax expression, out int constValue);
+        protected abstract bool IsIndexOfCall(TExpressionSyntax call, SemanticModel semanticModel);
+        protected abstract TExpressionSyntax Left(TBinaryExpressionSyntax binaryExpression);
+        protected abstract TExpressionSyntax Right(TBinaryExpressionSyntax binaryExpression);
+        protected abstract SyntaxToken OperatorToken(TBinaryExpressionSyntax binaryExpression);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+        protected DiagnosticDescriptor Rule { get; }
+
+        protected IndexOfCheckAgainstZeroBase(System.Resources.ResourceManager rspecResources) =>
+            Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecResources);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                GeneratedCodeRecognizer,
+                c =>
+                {
+                    var lessThan = (TBinaryExpressionSyntax)c.Node;
+                    if (TryGetConstantIntValue(Left(lessThan), out var constValue) &&
+                        constValue == 0 &&
+                        IsIndexOfCall(Right(lessThan), c.SemanticModel))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule,
+                            Left(lessThan).CreateLocation(OperatorToken(lessThan))));
+                    }
```

Comment:
This is a main logic of this rule and it repeats itself twice. Can you try to extract it into a method that will receive `Left(lessThan)`, `Right(lessThan)`?

The `.CreateLocation` sensitivity for argument order is bit annoying but can be fixed, as it's always the Token&FirstArgument. So you can swap them. I'd first try if `TextSpan.FromBounds` is sensitive to argument order :)
```suggestion
                    if (TryGetConstantIntValue(Left(lessThan), out var constValue) &&
                        constValue == 0 &&
                        IsIndexOfCall(Right(lessThan), c.SemanticModel))
                    {
                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule,
                            Left(lessThan).CreateLocation(OperatorToken(lessThan))));
                    }
```

Reply:
I extracted the condition but the if body did not as the FromBounds is sensitive to argument order. if the second argument is less then the first then It throws an ArgumentOutOfRangeException